Metrics/Micrometer guide

burtbeckwith opened the pull request.

burtbeckwith committed:
sort alphabetically

burtbeckwith committed:
metrics guide

burtbeckwith committed:
cleanup

burtbeckwith commented:
Preview: [metrics.pdf](https://github.com/micronaut-projects/micronaut-guides/files/8856335/metrics.pdf)

alvarosanchez commented on guides/micronaut-metrics/src/main/resources/application.yml:
```diff
@@ -0,0 +1,47 @@
+micronaut:
+  metrics:
+    enabled: true # <1>
+    export:
```
Should this line be commented out too?

alvarosanchez commented on guides/micronaut-metrics/java/src/main/java/example/micronaut/BookRepository.java:
```diff
@@ -0,0 +1,21 @@
+public interface BookRepository extends CrudRepository<Book, Long> { // <2>
+
+    @NonNull
+    List<Book> findAll();
```
Is this here because you prefer a `List<Book>` rather than an `Iterable<Book>`?

sdelamo committed:
add @GeneratedValue callout

sdelamo commented on guides/micronaut-metrics/groovy/src/main/groovy/example/micronaut/Book.groovy:
```diff
@@ -0,0 +1,26 @@
+class Book {
+
+    @Id // <2>
+    @GeneratedValue(AUTO)
```
it misses callout

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/542805a4180218cd482d80eb374d85898bae07d6

sdelamo commented on guides/micronaut-metrics/java/src/main/java/example/micronaut/Book.java:
```diff
@@ -0,0 +1,48 @@
+public class Book {
+
+    @Id // <2>
+    @GeneratedValue(AUTO)
```
It misses callout

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/542805a4180218cd482d80eb374d85898bae07d6

sdelamo commented on guides/micronaut-metrics/kotlin/src/main/kotlin/example/micronaut/Book.kt:
```diff
@@ -0,0 +1,13 @@
+@MappedEntity // <1>
+class Book(var isbn: String, var name: String) {
+    @Id // <2>
+    @GeneratedValue(AUTO)
```
It misses callout

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/542805a4180218cd482d80eb374d85898bae07d6

sdelamo commented on guides/micronaut-metrics/groovy/src/main/groovy/example/micronaut/BookRepository.groovy:
```diff
@@ -0,0 +1,19 @@
+    List<Book> findAll()
+
+    @NonNull
+    Optional<Book> findByIsbn(@NotNull String isbn)
```
Please use `@NotBlank` intead of `@NotNull` for String constraints.

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/8ac09a98937dbc3a3b5caa09baa031ca082fa912

sdelamo commented on guides/micronaut-metrics/java/src/main/java/example/micronaut/BookRepository.java:
```diff
@@ -0,0 +1,21 @@
+    List<Book> findAll();
+
+    @NonNull
+    Optional<Book> findByIsbn(@NotNull String isbn);
```
Please use `@NotBlank` intead of `@NotNull` for String constraints.

sdelamo replied:
resolved via resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/8ac09a98937dbc3a3b5caa09baa031ca082fa912

sdelamo commented on guides/micronaut-metrics/kotlin/src/main/kotlin/example/micronaut/BookRepository.kt:
```diff
@@ -0,0 +1,14 @@
+
+    override fun findAll(): List<Book>
+
+    fun findByIsbn(isbn: String): Optional<Book>
```
Add `@NotBlank` Constraint

sdelamo replied:
resolved via resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/8ac09a98937dbc3a3b5caa09baa031ca082fa912

sdelamo committed:
use MYSQL dialect not H2

sdelamo commented on guides/micronaut-metrics/groovy/src/main/groovy/example/micronaut/DataPopulator.groovy:
```diff
@@ -0,0 +1,33 @@
+import static io.micronaut.context.env.Environment.TEST
+
+@CompileStatic
+@Singleton
```
add callout

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/a32729d4a7a1108920bd68523310764576e7d8d7

sdelamo commented on guides/micronaut-metrics/java/src/main/java/example/micronaut/DataPopulator.java:
```diff
@@ -0,0 +1,31 @@
+
+import static io.micronaut.context.env.Environment.TEST;
+
+@Singleton
```
add callout

sdelamo replied:
resolved via resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/a32729d4a7a1108920bd68523310764576e7d8d7

sdelamo commented on guides/micronaut-metrics/kotlin/src/main/kotlin/example/micronaut/DataPopulator.kt:
```diff
@@ -0,0 +1,23 @@
+import jakarta.inject.Singleton
+import javax.transaction.Transactional
+
+@Singleton
```
add callout

sdelamo replied:
resolved via resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/a32729d4a7a1108920bd68523310764576e7d8d7

sdelamo commented on guides/micronaut-metrics/kotlin/src/main/kotlin/example/micronaut/DataPopulator.kt:
```diff
@@ -0,0 +1,23 @@
+import javax.transaction.Transactional
+
+@Singleton
+@Requires(notEnv = [TEST])
```
add callout

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/1f74943a608a7be2e1690114a29849619d881e2d

sdelamo commented on guides/micronaut-metrics/java/src/main/java/example/micronaut/DataPopulator.java:
```diff
@@ -0,0 +1,31 @@
+import static io.micronaut.context.env.Environment.TEST;
+
+@Singleton
+@Requires(notEnv = TEST)
```
add callout

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/1f74943a608a7be2e1690114a29849619d881e2d

sdelamo commented on guides/micronaut-metrics/groovy/src/main/groovy/example/micronaut/DataPopulator.groovy:
```diff
@@ -0,0 +1,33 @@
+
+@CompileStatic
+@Singleton
+@Requires(notEnv = TEST)
```
add callout

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/1f74943a608a7be2e1690114a29849619d881e2d

sdelamo commented on guides/micronaut-metrics/groovy/src/main/groovy/example/micronaut/DataPopulator.groovy:
```diff
@@ -0,0 +1,33 @@
+
+    private final BookRepository bookRepository
+
+    DataPopulator(BookRepository bookRepository) {
```
callout `constructor-di`

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/de4e9c716799bc341717953397560cc3a23017a6

sdelamo commented on guides/micronaut-metrics/java/src/main/java/example/micronaut/DataPopulator.java:
```diff
@@ -0,0 +1,31 @@
+
+    private final BookRepository bookRepository;
+
+    DataPopulator(BookRepository bookRepository) {
```
callout `constructor-di`

sdelamo replied:
resolved via resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/de4e9c716799bc341717953397560cc3a23017a6

sdelamo commented on guides/micronaut-metrics/kotlin/src/main/kotlin/example/micronaut/DataPopulator.kt:
```diff
@@ -0,0 +1,23 @@
+
+@Singleton
+@Requires(notEnv = [TEST])
+open class DataPopulator(private val bookRepository: BookRepository) {
```
callout `constructor-di`

sdelamo replied:
resolved via resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/de4e9c716799bc341717953397560cc3a23017a6

sdelamo commented on guides/micronaut-metrics/groovy/src/main/groovy/example/micronaut/DataPopulator.groovy:
```diff
@@ -0,0 +1,33 @@
+        this.bookRepository = bookRepository
+    }
+
+    @EventListener
```
add callout for `@EventListener`

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/7a952dfd349b2ec14983a02c2d7208d5ae3a6f11

sdelamo commented on guides/micronaut-metrics/java/src/main/java/example/micronaut/DataPopulator.java:
```diff
@@ -0,0 +1,31 @@
+        this.bookRepository = bookRepository;
+    }
+
+    @EventListener
```
add callout for `@EventListener`

sdelamo replied:
resolved via resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/7a952dfd349b2ec14983a02c2d7208d5ae3a6f11

sdelamo commented on guides/micronaut-metrics/kotlin/src/main/kotlin/example/micronaut/DataPopulator.kt:
```diff
@@ -0,0 +1,23 @@
+@Requires(notEnv = [TEST])
+open class DataPopulator(private val bookRepository: BookRepository) {
+
+    @EventListener
```
add callout for `@EventListener`

sdelamo replied:
resolved via resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/7a952dfd349b2ec14983a02c2d7208d5ae3a6f11

sdelamo committed:
singleton callout

sdelamo committed:
callout requires-not-env-test

sdelamo committed:
constructor-di callout

sdelamo committed:
event-listener pr

sdelamo commented on guides/micronaut-metrics/java/src/main/java/example/micronaut/DataPopulator.java:
```diff
@@ -0,0 +1,31 @@
+    }
+
+    @EventListener
+    @Transactional
```
Micronaut Data handles transaction demarcation automatically for you. Why are you using `@Transactional`? Are you adding the annotation to group the three inserts in a single transaction?

burtbeckwith replied:
you can make one comment that refers to 3 (or more) classes instead of spamming everyone. you can also commit more than one trivial change at a time instead of spamming everyone

sdelamo replied:
Sorry, I did separate commits in case you did not agree with any of them so that they could be easily reverted.

sdelamo commented on guides/micronaut-metrics/groovy/src/main/groovy/example/micronaut/DataPopulator.groovy:
```diff
@@ -0,0 +1,33 @@
+    }
+
+    @EventListener
+    @Transactional
```
Micronaut Data handles transaction demarcation automatically for you. Why are you using `@Transactional`? Are you adding the annotation to group the three inserts in a single transaction?

graemerocher replied:
No it doesn't not for multiple independent methods calls. There is nothing wrong with this code as written.

sdelamo commented on guides/micronaut-metrics/kotlin/src/main/kotlin/example/micronaut/DataPopulator.kt:
```diff
@@ -0,0 +1,23 @@
+open class DataPopulator(private val bookRepository: BookRepository) {
+
+    @EventListener
+    @Transactional
```
Micronaut Data handles transaction demarcation automatically for you. Why are you using `@Transactional`? Are you adding the annotation to group the three inserts in a single transaction?

sdelamo commented on guides/micronaut-metrics/kotlin/src/main/kotlin/example/micronaut/crypto/BitcoinPrice.kt:
```diff
@@ -0,0 +1,12 @@
+
+import io.micronaut.core.annotation.Introspected
+
+@Introspected
```
missing annotation

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/dedade8c8b61c274b272a6f45ebf5330dbc3e321

sdelamo commented on guides/micronaut-metrics/groovy/src/main/groovy/example/micronaut/crypto/BitcoinPrice.groovy:
```diff
@@ -0,0 +1,30 @@
+import io.micronaut.core.annotation.Introspected
+
+@CompileStatic
+@Introspected
```
missing callout

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/dedade8c8b61c274b272a6f45ebf5330dbc3e321

sdelamo commented on guides/micronaut-metrics/java/src/main/java/example/micronaut/crypto/BitcoinPrice.java:
```diff
@@ -0,0 +1,31 @@
+
+import io.micronaut.core.annotation.Introspected;
+
+@Introspected
```
missing callout

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/dedade8c8b61c274b272a6f45ebf5330dbc3e321

sdelamo committed:
add introspected callout

sdelamo commented on guides/micronaut-metrics/groovy/src/main/groovy/example/micronaut/crypto/PriceClient.groovy:
```diff
@@ -0,0 +1,11 @@
+@Client('https://api.kucoin.com/api/v1')
+interface PriceClient {
+
+    @Get('/market/orderbook/level1?symbol=BTC-USDT')
```
add a `@QueryValue` argument.

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/c6835ef4f90fffe5062d5fb1ab16a39c46eca574

sdelamo commented on guides/micronaut-metrics/groovy/src/main/groovy/example/micronaut/crypto/PriceClient.groovy:
```diff
@@ -0,0 +1,11 @@
+import io.micronaut.http.annotation.Get
+import io.micronaut.http.client.annotation.Client
+
+@Client('https://api.kucoin.com/api/v1')
```
Illustrate how to use use an `id` and supply the url via configuration. Makes easier to test an external dependency.

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/c6835ef4f90fffe5062d5fb1ab16a39c46eca574

sdelamo commented on guides/micronaut-metrics/java/src/main/java/example/micronaut/BookRepository.java:
```diff
@@ -0,0 +1,21 @@
+
+import static io.micronaut.data.model.query.builder.sql.Dialect.H2;
+
+@JdbcRepository(dialect = H2) // <1>
```
use MySQL dialect

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/8ac09a98937dbc3a3b5caa09baa031ca082fa912 and https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/1d1d63c09f7870eed66827431488148f4b0ff380

sdelamo commented on guides/micronaut-metrics/java/src/main/java/example/micronaut/crypto/CryptoService.java:
```diff
@@ -0,0 +1,44 @@
+@Singleton // <1>
+public class CryptoService {
+
+    private final Logger log = LoggerFactory.getLogger(getClass().getName());
```
rename `log` to `LOG`.

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/67e41d59e56c36f47021c61010b6d6e06a8817cb

sdelamo commented on guides/micronaut-metrics/kotlin/src/main/kotlin/example/micronaut/crypto/CryptoService.kt:
```diff
@@ -0,0 +1,40 @@
+    private val priceClient: PriceClient, // <2>
+    meterRegistry: MeterRegistry) {
+
+    private val log = LoggerFactory.getLogger(javaClass.name)
```
rename `log` to `LOG`

sdelamo replied:
resolved via https://github.com/micronaut-projects/micronaut-guides/pull/963/commits/67e41d59e56c36f47021c61010b6d6e06a8817cb

sdelamo committed:
rename log with LOG

sdelamo committed:
use mysql

sdelamo committed:
@QueryValue and @client  id

sdelamo committed:
Merge branch 'main' into metrics

sdelamo committed:
update to mysql

sdelamo committed:
swap order

burtbeckwith commented:
@sdelamo any plans to get the tests passing again?

sdelamo committed:
split test and mock http client call

sdelamo committed:
add mysql requirements

sdelamo committed:
better sections

sdelamo commented:
> @sdelamo any plans to get the tests passing again?

Yes, I am working on the guide right now. Sorry to break the build.

sdelamo committed:
add missing imports

sdelamo committed:
clarify copy

sdelamo committed:
write app application-test.yml

sdelamo committed:
clarify copy

sdelamo reviewed (approved):
It looks great to me, and this tutorial fills a significant gap (absence of monitoring material) that we had in guides: 

I have created several PRs with this guide as a base which I am happy to discuss.

- https://github.com/micronaut-projects/micronaut-guides/pull/972
- https://github.com/micronaut-projects/micronaut-guides/pull/971
- https://github.com/micronaut-projects/micronaut-guides/pull/969
- https://github.com/micronaut-projects/micronaut-guides/pull/968

kroegerk commented on guides/micronaut-metrics/metadata.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "title": "Collect Metrics with Micronaut",
+  "intro": "Learn how to collect standard and custom metrics with Micronaut.",
```
```suggestion
  "intro": "Learn how to collect standard and custom metrics with the Micronaut framework.",
```

kroegerk commented on guides/micronaut-metrics/micronaut-metrics.adoc:
```diff
@@ -0,0 +1,309 @@
+
+=== Metrics Configuration
+
+Several groups of metrics are enabled by default; these include system metrics (JVM info, uptime, etc.) as well as metrics tracking web requests, DataSource activity, and others. Overall metrics can be enabled or disabled, and groups can be individually enabled or disabled in configuration.
```
```suggestion
Several groups of metrics are enabled by default; these include system metrics (JVM info, uptime, etc.), as well as metrics tracking web requests, DataSource activity, and others. Overall metrics can be enabled or disabled, and groups can be individually enabled or disabled in configuration.
```

kroegerk commented on guides/micronaut-metrics/micronaut-metrics.adoc:
```diff
@@ -0,0 +1,309 @@
+
+source:BookRepository[]
+callout:jdbcrepository[1]
+<2> Specifies that `Book` is the root entity and the primary key type is `Long`.
```
```suggestion
<2> Specifies that `Book` is the root entity, and the primary key type is `Long`.
```

kroegerk commented on guides/micronaut-metrics/micronaut-metrics.adoc:
```diff
@@ -0,0 +1,309 @@
+
+==== Service
+
+Create the `CryptoService` class which uses `PriceClient` and updates three custom meters:
```
```suggestion
Create a `CryptoService` class that uses `PriceClient` and updates three custom meters:
```

kroegerk commented on guides/micronaut-metrics/micronaut-metrics.adoc:
```diff
@@ -0,0 +1,309 @@
+
+testResource:application-test.yml[]
+
+<1> Configure the default datasource to use the MySQL database provided by TestContainers for the test environment. By using a specially modified JDBC URL, Testcontainers provides a disposable stand-in database that can be used without requiring modification to your application code. See https://www.testcontainers.org/modules/databases/jdbc/[Test Containers JDBC URL].
```
Should Testcontainers have the "C" capitalized? It looks like only the "T" should be capitalized. 

https://github.com/testcontainers/testcontainers-java 

Whichever one you pick, I'd do a quick search and make sure you're consistent throughout.

shingjo commented:
My 2 cents:

#971

The callout text is meaningless:
 
 > You can declare a method as transactional with the `javax.transaction.Transactional` annotation.

Does not explain what Transactional does or why to use it. Nor does it link to our own documentation so that a novice could learn more about it.

---

#969

H2 was a better choice, the focus of the guide is Metrics. It allows the reader to use the guide as a simple test bed. A note or callout about using H2 in production would suffice. 

---

#968

If Atlas is not going to be explained and shown, it should be removed.  There should be text about the different Registries supported by Micronaut. Namely because with Starter you add the metrics dependencies based on what Registry you are going to use.

---

The guide does not introduce and show the `@Counted` or `@Timed` annotations. (Aren't annotations the preferred way?)

With that said, demonstrating the use of the MeterRegistry is great and should be included. (Do think there should be more text about `gauge`, reading the guide a reader might infer that any type object could be stored.)

{Personal note: For me guide lacks continuity (Books, Crypto, Books-Flyway, Crypto-Running) that I initially missed out on all the good information in the tests (lazy loading of metrics and the use of MeterRegistry Tags; after looking up what they did.}

graemerocher commented on guides/micronaut-metrics/java/src/main/java/example/micronaut/crypto/CryptoService.java:
```diff
@@ -0,0 +1,44 @@
+
+    @Scheduled(fixedRate = "${crypto.updateFrequency:1h}",
+            initialDelay = "${crypto.initialDelay:0s}") // <6>
+    public void updatePrice() {
```
As per what John said the use of `timer` can be replaced with `@Timed` and `@Counted` on the method

sdelamo replied:
@shingjo can you provide a PR against this guide showing the usage of `@Timed` and `@Counted` ?

shingjo replied:
@sdelamo Can do.

sdelamo committed:
revert the guide to use H2 (#969)

I prefer to avoid the usage in H2 in guides. I would like for use not to produce any documentation which uses H2.

I have seen people using H2 for test and PostgresSQL in Micronaut even if we have made it difficutl. I don't want us to encourage that behaviour.

But if you feel strongly the metrics guide is better using H2. This reverts my change to use MySQL.

sdelamo committed:
transactional callout (#971)

sdelamo committed:
remove atlas mention (#968)

Co-authored-by: Burt Beckwith <burt@burtbeckwith.com>

burtbeckwith committed:
fixed the "fixes"
misc cleanup

burtbeckwith committed:
Merge branch 'master' into metrics

# Conflicts:
#	src/docs/common/callouts/callout-client.adoc
#	src/docs/common/snippets/common-flyway.adoc

burtbeckwith committed:
a few more minor updates

burtbeckwith committed:
Update guides/micronaut-metrics/metadata.json

Co-authored-by: Kristin Kroeger <33693760+kroegerk@users.noreply.github.com>

burtbeckwith committed:
Update guides/micronaut-metrics/micronaut-metrics.adoc

Co-authored-by: Kristin Kroeger <33693760+kroegerk@users.noreply.github.com>

burtbeckwith committed:
Update guides/micronaut-metrics/micronaut-metrics.adoc

Co-authored-by: Kristin Kroeger <33693760+kroegerk@users.noreply.github.com>

burtbeckwith committed:
Update guides/micronaut-metrics/micronaut-metrics.adoc

Co-authored-by: Kristin Kroeger <33693760+kroegerk@users.noreply.github.com>

timyates committed:
Add screencasts with reactive CSS (#1014)

burtbeckwith committed:
add author (#1018)

dependabot[bot] committed:
Bump selenium-chrome-driver in /buildSrc/src/main/resources (#1004)

Bumps [selenium-chrome-driver](https://github.com/SeleniumHQ/selenium) from 4.2.2 to 4.3.0.
- [Release notes](https://github.com/SeleniumHQ/selenium/releases)
- [Commits](https://github.com/SeleniumHQ/selenium/commits/selenium-4.3.0)

---
updated-dependencies:
- dependency-name: org.seleniumhq.selenium:selenium-chrome-driver
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>

Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

sdelamo committed:
build: update to Micronaut 3.5.3 (#1016)

graemerocher commented:
There is still no example in this guide showing how to use `@Counted` and `@Timed`

graemerocher commented:
perhaps adding the annotations to a method like this:

https://github.com/micronaut-projects/micronaut-guides/pull/963/files#diff-695775e144eff330ac10c64129ba02dce2b3078348e729135b6510f1fc311f6bR12

and showing you can time and count repository calls with `@Counted` and `@Timed` would suffice

Commit:
Added @Timed and @Counted to the guide

jeffscottbrown committed:
fix minor grammar issue (#1024)

jeffscottbrown committed:
minor grammar correction (#1023)

burtbeckwith committed:
native image -> native executable (#1015)

sdelamo committed:
build: Micronaut 3.5.4 (#1028)

sdelamo committed:
Set publication data to Aug 1

sdelamo committed:
twitter card
